workflow-core/src/main/java/com/squareup/workflow/WorkflowOperators.kt

```diff
@@ -73,10 +76,36 @@ fun <S1 : Any, S2 : Any, E : Any, O : Any> Workflow<S1, E, O>.switchMapState(
     WorkflowInput<E> by this {
   override fun openSubscriptionToState(): ReceiveChannel<S2> =
     GlobalScope.produce(operatorContext, capacity = CONFLATED) {
-      val source = this@switchMapState.openSubscriptionToState()
-      source.consumeEach { rawState ->
-        transform(rawState).consumeEach { transformedState ->
-          send(transformedState)
+      val upstreamChannel = this@switchMapState.openSubscriptionToState()
+      var transformedChannel: ReceiveChannel<S2>? = null
+      val downstreamChannel = channel
+
+      upstreamChannel.consume {
+        whileSelect {
+          upstreamChannel.onReceiveOrNull { upstreamState ->
+            if (upstreamState == null) {
```

Comment:
I wonder if both of these would be more readable flipped and converted to use `?.let / ?:`.

Reply:
That feels like code golf to me. `let` improves readability when the nullable value is being transformed, in this case it's not, it's being used as a sentinel for a completely different signal (channel close). This isn't ideal obviously and there's an open issue to make a more explicit receive call that doesn't use null.